### PR TITLE
Add router.off method support filter

### LINKED ISSUE #247: router.off method support host filter
When there are two identical paths in the route, but the host is different

```js
router.on('GET', '/a', { constraints: { host: '1' } }, () => {})
router.on('GET', '/a', { constraints: { host: '2' } }, () => {})
```

When executing the off method, want to filter by host, like:

```js
router.off('GET', '/a', {
  host: '1'
})
```

But now only the path parameter is supported，Run the following code and both routes will be deleted

```js
router.off('GET', '/a')

router.prettyPrint() // is empty
```

### PR DESCRIPTION
**Pull request by OrangeXC**
fix #247

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -320,23 +320,30 @@ Router.prototype.off = function off (method, path) {
...
+    this.off(method, pathFull, derivedConstraints)
+    this.off(method, pathOptional, derivedConstraints)
     return
   }
 
+  function matcher (currentConstraints = {}) {
+    if (!derivedConstraints) return true
+
+    return Object.keys(derivedConstraints)
+      .every((key) => derivedConstraints[key] === currentConstraints[key])
```
Comment (bugagashenkj) [inline index.js:332]: Why do you check only keys and not check values?

Comment (bugagashenkj) [inline index.js:332]: You should use deepEqual like here to compare constraints.
https://github.com/delvedor/find-my-way/blob/b38003a06b59da0ba336522babb933d465db3b96/custom_node.js#L207

File: index.js [hunk 2]
```diff
@@ -320,23 +320,30 @@ Router.prototype.off = function off (method, path) {
...
+  }
+
   // Rebuild tree without the specific route
   const ignoreTrailingSlash = this.ignoreTrailingSlash
   var newRoutes = self.routes.filter(function (route) {
     if (!ignoreTrailingSlash) {
-      return !(method === route.method && path === route.path)
+      return !(method === route.method && path === route.path && matcher(route.opts?.constraints))
```
Comment (bugagashenkj) [inline index.js:339]: The optional chaining operator only appeared in nodejs 14, we can't use it since we support the 12 version.

**Review by bugagashenkj — CHANGES_REQUESTED** [review]

**Review by bugagashenkj — APPROVED** [review]
LGTM

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you document this in the README?

**Comment by OrangeXC** [comment]
> Can you document this in the README?

Done

File: README.md [hunk 3]
```diff
@@ -383,6 +383,14 @@ router.off(['GET', 'POST'], '/example')
 // => null
 ```
 
+##### off(methods, path, [constraints])
+Deregister a route for each `constraints` key is matched, containing keys like the `host` for the request, the `version` for the route to be matched, or other custom constraint values.
```
Comment (mcollina) [inline README.md:387]:
```suggestion
Deregister a route for each `constraints` key is matched, containing keys like the `host` for the request, the `version` for the route to be matched, or other custom constraint values. See the [constraints section](https://github.com/delvedor/find-my-way#constraints) to know more.
```

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
Still lgtm

**Comment by OrangeXC** [comment]
@mcollina @bugagashenkj   

Can this PR be merged?

**Comment by mcollina** [comment]
Unless @delvedor beats me to it, I've limited bandwidth to ship this in a release before May.

**Review by Eomm — APPROVED** [review]

**Pull request merged**